### PR TITLE
Detect icmpv6hdr v4

### PR DESCRIPTION
**Pull request by catenacyber**
Link to [redmine](https://redmine.openinfosecfoundation.org/projects/suricata/issues) ticket:
https://redmine.openinfosecfoundation.org/issues/3323 

Describe changes:
- Adds `icmpv6.hdr` signature keyword
- Fixes typos in comments
- Adds missing doc for `ipv6.hdr` keyword
- Adds a generic `DetectU32Data` structure for signature matches on uint32 values
- Adds `icmpv6.mtu` signature  keyword
- Fix endianness for ICMPv6 MTU decoding

This fixes only the first evasion of the redmine ticket

Makes https://github.com/OISF/suricata-verify/pull/187 pass

Modifies #4631 by fixing use of `SCReturnPtr`

There is still some points to improve/discuss :
- The prefilter magic with my dummy `du32.arg2 = UINT32_MAX;`
- Should `flags` be updated even if the function `DetectICMPv6hdrSetup` fails?
- Handle `DetectBufferTypeRegister` as fatal error... Should it be part of this PR ?
- Use `pcre_copy_substring` in detect-tcpmss.c... Should it be part of this PR ?

### CODE REVIEW

File: src/detect-engine-prefilter-common.h [hunk 1]
```diff
@@ -94,4 +94,14 @@ PrefilterPacketHeaderExtraMatch(const PrefilterPacketHeaderCtx *ctx,
     return TRUE;
 }
 
+static inline bool PrefilterIsPrefilterableById(const Signature *s, enum DetectKeywordId kid) {
+    const SigMatch *sm;
+    for (sm = s->init_data->smlists[DETECT_SM_LIST_MATCH] ; sm != NULL; sm = sm->next) {
+        if (sm->type == kid) {
+            return TRUE;
```
Comment (jlucovsky) [inline src/detect-engine-prefilter-common.h]: Use `true` and `false` for `bool` types.

File: src/detect-util-uint.c [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+ * \param u32str Pointer to the user provided u32 options
+ *
+ * \retval DetectU32Data pointer to DetectU32Data on success
+ * \retval NULL on failure
+ */
+
+DetectU32Data *DetectU32Parse (const char *u32str)
+{
+    DetectU32Data *u32d = NULL;
+    char arg1[16] = "";
```
Comment (jlucovsky) [inline src/detect-util-uint.c]: I don't think these need to be initialized.

File: src/detect-util-uint.c [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+    // not enough place for a->arg2 in PrefilterPacketHeaderValue
+}
+
+bool
+PrefilterPacketU32Compare(PrefilterPacketHeaderValue v, void *smctx)
+{
+    const DetectU32Data *a = smctx;
+    if (v.u8[0] == a->mode &&
+        v.u32[1] == a->arg1)
+        return TRUE;
```
Comment (jlucovsky) [inline src/detect-util-uint.c]: `true` and `false` with `bool`

File: src/detect-engine-prefilter-common.h (hunk 1 above)
Comment (catenacyber) [reply src/detect-engine-prefilter-common.h]: ok

File: src/detect-util-uint.c (hunk 2 above)
Comment (catenacyber) [reply src/detect-util-uint.c]: Are you sure `pcre_copy_substring` does the empty string initialization ?

Other question, what is the right size for these buffers ?

File: src/detect-util-uint.c (hunk 3 above)
Comment (catenacyber) [reply src/detect-util-uint.c]: ok

File: src/detect-engine-prefilter-common.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-engine-prefilter-common.h]: Ok, and fixed in detect-tcpmss.c where this code was copied

File: src/detect-icmpv6-mtu.c [hunk 4]
```diff
@@ -0,0 +1,186 @@
...
+#endif
+    sigmatch_table[DETECT_ICMPV6MTU].SupportsPrefilter = PrefilterIcmpv6mtuIsPrefilterable;
+    sigmatch_table[DETECT_ICMPV6MTU].SetupPrefilter = PrefilterSetupIcmpv6mtu;
+
+    DetectU32Register();
+    return;
+}
+
+// returns 0 on no mtu, and 1 if mtu
+static inline int packetGetIcmpv6MTU(Packet *p, uint32_t *picmpv6mtu) {
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-icmpv6-mtu.c]: ok for new line.
What is the function name style ?

File: src/detect-icmpv6-mtu.c [hunk 5]
```diff
@@ -0,0 +1,186 @@
...
+    if (!(PKT_IS_ICMPV6(p)) || PKT_IS_PSEUDOPKT(p))
+        return 0;
+    if (!(ICMPV6_HAS_MTU(p)))
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-icmpv6-mtu.c]: This should be done by `PKT_IS_ICMPV6(p)`

But I missed the condition `ICMPV6_GET_CODE(p) != 0`

Nice review :-)

File: src/detect-util-uint.c [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+    return FALSE;
+}
+
+static int detectu32Registered = 0;
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-util-uint.c]: like `detect_u32_registered`?

File: src/detect-util-uint.c [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+    return NULL;
+}
+
+void
+PrefilterPacketU32Set(PrefilterPacketHeaderValue *v, void *smctx)
+{
+    const DetectU32Data *a = smctx;
+    v->u8[0] = a->mode;
+    v->u32[1] = a->arg1;
+    // not enough place for a->arg2 in PrefilterPacketHeaderValue
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-util-uint.c]: Ok, I will propose a solution where we accept not supporting it

File: src/detect-icmpv6-mtu.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-icmpv6-mtu.c]: https://redmine.openinfosecfoundation.org/projects/suricata/wiki/Coding_Style

File: src/detect-util-uint.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-util-uint.c]: Why not extend the side of the storage? It seems we can do it for free.

File: src/detect-util-uint.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-util-uint.c]: I generally prefix with `g_` when its at the root of a C file, but I don't think this is done consistently.

File: src/detect-icmpv6-mtu.c [hunk 8]
```diff
@@ -0,0 +1,186 @@
...
+{
+    return PrefilterSetupPacketHeader(de_ctx, sgh, DETECT_ICMPV6MTU,
+            PrefilterPacketU32Set,
+            PrefilterPacketU32Compare,
+            PrefilterPacketIcmpv6mtuMatch);
+}
+
+static bool PrefilterIcmpv6mtuIsPrefilterable(const Signature *s)
+{
+    return PrefilterIsPrefilterableById(s, DETECT_ICMPV6MTU);
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-icmpv6-mtu.c]: Modifying this and adding some TODO comment

File: src/detect-util-uint.c [hunk 9]
```diff
@@ -0,0 +1,232 @@
+/* Copyright (C) 2020 Open Information Security Foundation
```
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-util-uint.c]: ok

File: src/detect-util-uint.c (hunk 6 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-util-uint.c]: ok

File: src/detect-icmpv6-mtu.c (hunk 4 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-icmpv6-mtu.c]: Ok for `DetectICMPv6mtuGetValue` ?

File: src/detect-util-uint.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-util-uint.c]: I may be too careful but this code looks to me like something optimized whatever the ABI of the system... Are you sure ?

**Comment by victorjulien** [comment]
Partly merged through #4637. Not merged: mtu patches and tcpmss patch.

File: src/detect-util-uint.c (hunk 7 above)
[the remark this answers is not in the document]
Comment (victorjulien) [reply src/detect-util-uint.c]: So in master we have:
```c
struct PrefilterPacketHeaderCtx_ {
	PrefilterPacketHeaderValue v1;                   /*     0     8 */
	uint16_t                   type;                 /*     8     2 */
	uint16_t                   value;                /*    10     2 */
	uint32_t                   sigs_cnt;             /*    12     4 */
	uint32_t *                 sigs_array;           /*    16     8 */

	/* size: 24, cachelines: 1, members: 5 */
	/* last cacheline: 24 bytes */
};
```
If we add a u64:
```diff
diff --git a/src/detect-engine-prefilter-common.c b/src/detect-engine-prefilter-common.c
index 4c9e4f5e0..85af66081 100644
--- a/src/detect-engine-prefilter-common.c
+++ b/src/detect-engine-prefilter-common.c
@@ -31,7 +31,7 @@ typedef struct PrefilterPacketHeaderHashCtx_ {
 static uint32_t PrefilterPacketHeaderHashFunc(HashListTable *ht, void *data, uint16_t datalen)
 {
     PrefilterPacketHeaderCtx *ctx = data;
-    uint64_t hash = ctx->v1.u64 + ctx->type + ctx->value;
+    uint64_t hash = ctx->v1.u64[0] + ctx->v1.u64[1] + ctx->type + ctx->value;
     hash %= ht->array_size;
     return hash;
 }
@@ -41,7 +41,8 @@ static char PrefilterPacketHeaderCompareFunc(void *data1, uint16_t len1,
 {
     PrefilterPacketHeaderHashCtx *ctx1 = data1;
     PrefilterPacketHeaderHashCtx *ctx2 = data2;
-    return (ctx1->v1.u64 == ctx2->v1.u64 &&
+    return (ctx1->v1.u64[0] == ctx2->v1.u64[0] &&
+           ctx1->v1.u64[1] == ctx2->v1.u64[1] &&
             ctx1->type == ctx2->type &&
             ctx1->value == ctx2->value);
 }
diff --git a/src/detect-engine-prefilter-common.h b/src/detect-engine-prefilter-common.h
index b8f317f14..b2bd461e2 100644
--- a/src/detect-engine-prefilter-common.h
+++ b/src/detect-engine-prefilter-common.h
@@ -19,10 +19,10 @@
 #define __DETECT_ENGINE_PREFILTER_COMMON_H__
 
 typedef union {
-    uint8_t u8[8];
-    uint16_t u16[4];
-    uint32_t u32[2];
-    uint64_t u64;
+    uint8_t u8[16];
+    uint16_t u16[8];
+    uint32_t u32[4];
+    uint64_t u64[2];
 } PrefilterPacketHeaderValue;
 
 #define PREFILTER_EXTRA_MATCH_UNUSED  0
```
We get:
```c
struct PrefilterPacketHeaderCtx_ {
	PrefilterPacketHeaderValue v1;                   /*     0    16 */
	uint16_t                   type;                 /*    16     2 */
	uint16_t                   value;                /*    18     2 */
	uint32_t                   sigs_cnt;             /*    20     4 */
	uint32_t *                 sigs_array;           /*    24     8 */

	/* size: 32, cachelines: 1, members: 5 */
	/* last cacheline: 32 bytes */
};
```
Where do you see the potential for side effects of this?

[the remark this answers is not in the document]
Comment (catenacyber) [reply src/detect-util-uint.c]: _I_ do not see side effects... But I was wary for performance issues...
You look confident.
I will propose this version

**Pull request closed without merging**